### PR TITLE
[6X] pg_dump and pg_upgrade: fix/change external partition handling

### PR DESCRIPTION
**Pull request by jimmyyih**
The individual commits that make upgrading external partitions work:
```
commit 1495eec10b5bc1485e7b1eaaf9f6c2a4897ce175
Author: Jimmy Yih <jyih@vmware.com>
Date:   Thu Jan 19 18:42:13 2023 -0800

    Allow EXCHANGE PARTITION to run in utility mode during upgrade

    We don't allow EXCHANGE PARTITION to run in utility mode because
    primary segments do not have the pg_partition_rule catalog information
    to be able to perform the EXCHANGE PARTITION. However, we should allow
    it to run during upgrade because only the coordinator segment is
    touched which does have the pg_partition_rule data (its catalog is
    later copied over to the primary segments as a part of the gpupgrade
    flow).

commit 0e853d6e20933d8d06c961a0181decf651f46e47
Author: Jimmy Yih <jyih@vmware.com>
Date:   Fri Jan 20 14:32:43 2023 -0800

    Remove check for external partitions

    In an old GPDB commit, external partitions were banned from being
    upgraded because ALTER TABLE EXCHANGE PARTITION was banned in utility
    mode. The ALTER TABLE EXCHANGE PARTITION code could have been modified
    to work with upgrade but the logic would not work when primary
    segments were upgraded due to their lack of pg_partition_rule
    data. Nowadays however, we only upgrade the coordinator segment and
    copy its catalog to the target primary segments as a part of gpupgrade
    logic. Because of that, we can remove the check for external
    partitions and fix the underlying ALTER TABLE EXCHANGE PARTITION issue
    instead.

    Old GPDB commit reference:
    https://github.com/greenplum-db/gpdb/commit/ceda2890a43247a4b97ee68b9970240cb04977d0

commit 3ee14a583de20f842b387c5c931bf00cf145ffa0
Author: Jimmy Yih <jyih@vmware.com>
Date:   Tue Jan 24 16:58:53 2023 -0800

    Fix pg_dump external partition logic for multi-level partition tables

    The old logic was only able to handle simple partition tables that did
    not contain subpartitions. Also, the query being used was a bit
    excessive for the result produced. To fix the issue, we now do two
    simple queries: one query that gets the external partitions of a
    partition table and another query that gets the subroot hierarchy to
    be used when dealing with a multi-level partition table. Using the two
    query results and some result parsing, we now properly output the
    ALTER TABLE EXCHANGE PARTITION to contain the required ALTER PARTITION
    syntax for each subroot.

    Fixes https://github.com/greenplum-db/gpdb/issues/5003.

    Co-authored-by: Kalen Krempely <kkrempely@vmware.com>

commit 3a0d211f8a500c6bc15fcc98fb39c4c2db704eae
Author: Jimmy Yih <jyih@vmware.com>
Date:   Tue Jan 24 17:24:56 2023 -0800

    Fix pg_dump temporary external table logic for external partitions

    When dealing with external partitions, pg_dump outputs SQL to create a
    temporary external table that will later be used in an ALTER TABLE
    EXCHANGE PARTITION. This worked fine for most pg_dump cases but did
    not work for binary upgrade scenarios where the temporary external
    table name generated was longer than NAMEDATALEN. During restore, the
    temporary external table was not able to obtain the preserved OID. The
    preserved OID is mapped to the temporary external table's full name
    but CREATE EXTERNAL TABLE truncates the name down to NAMEDATALEN which
    results in the temporary external table getting a new OID instead of
    using the preserved OID. Afterwards, pg_upgrade would error with an
    OID mismatch failure.

    To fix the issue, the logic to generate the temporary external table
    name has been changed to always be under length NAMEDATALEN while
    staying unique and identifiable to the partition. With that, the
    preserved OID will always be used during restore.
```

Test coverage PR in the gpupgrade repository:
https://github.com/greenplum-db/gpupgrade/pull/747

### CODE REVIEW

File: contrib/pg_upgrade/greenplum/check_gp.c [hunk 1]
```diff
@@ -17,9 +17,6 @@
 #include "pg_upgrade_greenplum.h"
 #include "check_gp.h"
 
-#define RELSTORAGE_EXTERNAL	'x'
-
-static void check_external_partition(void);
```
Comment (kalensk) [inline contrib/pg_upgrade/greenplum/check_gp.c:22]: We should make a note to updated the public facing gpupgrade docs:
https://docs.vmware.com/en/VMware-Tanzu-Greenplum-Upgrade/1.5/greenplum-upgrade/GUID-gpupgrade_initialize_checks.html#external-tables-used-in-partitioning-6

File: src/bin/pg_dump/pg_dump.c [hunk 2]
```diff
@@ -13769,87 +13776,190 @@ dumpTableSchema(Archive *fout, TableInfo *tbinfo)
...
 			{
-				char tmpExtTable[500] = {0};
-				relname = pg_strdup(PQgetvalue(res, i, i_relname));
-				snprintf(tmpExtTable, sizeof(tmpExtTable), "%s%s", relname, EXT_PARTITION_NAME_POSTFIX);
-				char *qualTmpExtTable = pg_strdup(fmtQualifiedId(fout->remoteVersion,
-																 tbinfo->dobj.namespace->dobj.name,
-																 tmpExtTable));
-
+				int parlevel = atoi(PQgetvalue(getExternalPartsResult, i, i_parlevel));
+				int subRootPartitions[parlevel];
```
[the remark this answers is not in the document]
Comment (jimmyyih) [reply src/bin/pg_dump/pg_dump.c]: Updated.

File: src/bin/pg_dump/pg_dump.c [hunk 3]
```diff
@@ -13769,87 +13776,190 @@ dumpTableSchema(Archive *fout, TableInfo *tbinfo)
...
+				}
+
+				/*
+				 * Add all the ALTER PARTITION info needed for the ALTER TABLE
+				 * EXCHANGE PARTITION statement. The subRootPartitions array
+				 * contains the subroot partition hierarchy info indexed
+				 * sequentially by the partition level already. If it is an
+				 * anonymous range partition we must exchange for the rank
+				 * rather than the parname.
+				 */
```
[the remark this answers is not in the document]
Comment (jimmyyih) [reply src/bin/pg_dump/pg_dump.c]: I redid the comment to be more clear using your suggestion as the main point that we're trying to convey.

File: contrib/pg_upgrade/greenplum/check_gp.c (hunk 1 above)
Comment (jimmyyih) [reply contrib/pg_upgrade/greenplum/check_gp.c:22]: Thanks; I'll make note of this.

File: src/bin/pg_dump/pg_dump.c [hunk 4]
```diff
@@ -4972,11 +4970,20 @@ getTables(Archive *fout, int *numTables)
...
+			 * below will generate a temporary table name
+			 * _external_partition_parrelid_<oid>_relid_<oid> which is under
+			 * length NAMEDATALEN and is unique. This generated name will be
+			 * stored in the table map and used later to assist in the ALTER
+			 * TABLE EXCHANGE PARTITION dump logic.
 			 */
-			char tmpStr[500];
-			snprintf(tmpStr, sizeof(tmpStr), "%s%s", tblinfo[i].dobj.name, EXT_PARTITION_NAME_POSTFIX);
+			char tmpStr[NAMEDATALEN];
+			snprintf(tmpStr, NAMEDATALEN, "_external_partition_parrelid_%u_relid_%u", tblinfo[i].parrelid, tblinfo[i].dobj.catId.oid);
```
Comment (kalensk) [inline src/bin/pg_dump/pg_dump.c:4986]: > This is especially important for binary upgrade dumps where the preserved oid expects a certain relation name that is implicitly restricted to NAMEDATALEN.

Thank you for adding this. Super useful to have here!

**Comment by kalensk** [comment]
Im assuming no regress or isolation type tests are needed? Just pg_upgrade acceptance tests.

**Review by kalensk — APPROVED** [review]

**Comment by jimmyyih** [comment]
> Im assuming no regress or isolation type tests are needed? Just pg_upgrade acceptance tests.

Yeah, the pg_upgrade acceptance tests are fine.  It's more ideal to test this scenario there anyways.

**Pull request merged**